Item Pickups and Inventory Interactions

kylegessel opened the pull request:
This pull request fortifies our earlier implementation of item inventory and cleans up the player pickup class by giving each item their own pickup class.

Final remaining items are also added. Bomb, boomerang, and Blue Rupees have been added to the entities list and can now be found as items in the first room.

Link now responds to changes in his inventory and checks his inventory before executing commands.
- Link no longer uses the thrown sword if he has less than full health.
- Link no longer uses a bomb if has none remaining in his inventory.
- Link can only use the boomerang after he has picked it up for the first time.

There was also some tweaking of the sword hit-box. It was not being used previously, but now can be used to kill enemies. Its class still needs some attention and its one of the classes I really want to rework.

HUD items (compass and map) as well as the clock still need method bodies.

kylegessel commented:
Newest push added new items to the dictionary, and cleaned up some the the pickup classes.

Boomerang can now only be found as a pickup in room 11.

Arrows won't fire unless the link has picked up the bow!

Bailey-Ardrey commented on Sprintfinity3902/Entities/Items/SwordHitboxItem.cs:
```diff
@@ -62,7 +62,7 @@ public override void Update(GameTime gameTime)
         public void MoveItem()
         {
-            if (itemUseCount <= 5) //this time amount needs to be tweaked.
+            if (itemUseCount <= 30) //this time amount needs to be tweaked.
```
At some point this Magic Number might need to be taken out.

kylegessel replied:
Thanks for catching this, just gave it a constant.

Bailey-Ardrey commented on Sprintfinity3902/States/ItemPickup/FairyPickup.cs:
```diff
@@ -0,0 +1,24 @@
+            Link.LinkHealth = Link.MaxHealth;
+            Link.heartChanged = true;
+
+            Sound.SoundLoader.Instance.GetSound(Sound.SoundLoader.Sounds.LOZ_Get_Heart).Play(Global.Var.VOLUME, Global.Var.PITCH, Global.Var.PAN);
```
I am not sure if the Fairy has a specific sound associated.

kylegessel replied:
Removing it now!

Bailey-Ardrey commented on Sprintfinity3902/States/ItemPickup/HeartContainerPickup.cs:
```diff
@@ -0,0 +1,25 @@
+            Link.MaxHealth += 2;
+            Link.LinkHealth += 2;
+            Link.heartChanged = true;
+            Sound.SoundLoader.Instance.GetSound(Sound.SoundLoader.Sounds.LOZ_Get_Heart).Play(Global.Var.VOLUME, Global.Var.PITCH, Global.Var.PAN);
```
I am not sure if the Heart Container has a specific sound associated.

kylegessel replied:
Removed it!

Bailey-Ardrey commented on Sprintfinity3902/States/ItemPickup/HeartPickup.cs:
```diff
@@ -0,0 +1,32 @@
+                    Link.LinkHealth++;
+                }
+                Link.heartChanged = true;
+                Sound.SoundLoader.Instance.GetSound(Sound.SoundLoader.Sounds.LOZ_Get_Heart).Play(Global.Var.VOLUME, Global.Var.PITCH, Global.Var.PAN);
```
Heart pickup sound is not working at runtime for me. Are you having the same issue?

kylegessel replied:
Yup, this is a bug. Heart sound only plays if link is damaged.

Updating so it plays on every pickup.

Bailey-Ardrey reviewed (changes requested):
Everything looks good really good! Pickup cleans up a lot in player. Only two things I noticed:

1. The heart sound isn't playing when Link picks up a heart.
2. An error is thrown when Link dies in this iteration of the code... Game Over state is broken.

Bailey-Ardrey commented:
It actually looks like Game Over was also broken in my pr too.

Bailey-Ardrey commented:
It has to do with the Sound Manager being Reset in the Win and Lose Wrapper. I think we can ignore it for now, as it was broken before this pr.

kylegessel commented:
Added the changes requested!
 
- Some items had their sound removed (Fairy and Heart Container). 
- Heart sound plays when link picks up a heart, even while full health.
- Removed a magic number.

Bailey-Ardrey reviewed (approved):
Looks good! Ready to merge. I'll let Casey get his comments in first.

caseyvanc commented:
This looks good to me. I like the item pickup implementation.
